Fees

imprfekt opened the pull request.

imprfekt committed:
Proposal for implementing fees

Adding a percentage fee, and an idea for implementation of partner fee and a partner-specific fee override (effectively enabling a discount or an upsell)

Commit:
added utils.Participants enum for action permission resolving, added calls in cancel and transfer methods, wip, needs tests and proper naming, should move from util probably

Commit:
fix typo and phantom log

parazyd committed:
cargo fmt

imprfekt committed:
cargo fmt

imprfekt committed:
enum rename (Participant => Invoker)

Commit:
split testing fascilities in different module;added tests for cancel and transfer control checks, removed unecessary check in transfer validation code

parazyd committed:
Refactor project into multiple files and clean up.

Commit:
remove transfer recipient validation

Commit:
merge remote changes

Commit:
remove commented assertion in cancel test

parazyd committed:
safety: Check data is empty for uninitialized account check.

nemanjab17 commented on src/cancel_stream.rs:
```diff
@@ -0,0 +1,293 @@
+
+    // Sanity checks
+    initialized_account_sanity_check(program_id, acc.clone())?;
+    metadata_sanity_check(acc.clone())?;
```
can we borrow here?

nemanjab17 commented on src/cancel_stream.rs:
```diff
@@ -0,0 +1,293 @@
+
+    let recipient_available = calculate_available(
+        now,
+        metadata.ix.clone(),
```
maybe borrow here too

nemanjab17 commented on src/cancel_stream.rs:
```diff
@@ -0,0 +1,293 @@
+        metadata.streamflow_fee_withdrawn,
+    );
+
+    let partner_available = calculate_available(
```
maybe move calculate_available to be a member of TokenStreamData since all the data needed to calculate it is member of TokenStreamData already

nemanjab17 commented on src/cancel_stream.rs:
```diff
@@ -0,0 +1,293 @@
+    let seeds = [acc.metadata.key.as_ref(), &[escrow_tokens_bump]];
+
+    if recipient_available > 0 {
+        msg!("Transferring unlocked tokens to recipient");
```
This block is repeated for each participant token account, and both in cancel stream and withdraw stream methods, any future additions would need multiple updates all over the code base, is there a cheap way of making it easier for maint?

imprfekt commented on src/cancel_stream.rs:
```diff
@@ -0,0 +1,293 @@
+/// if there are any unlocked funds. If so, they will be transferred to the
+/// stream recipient.
+pub(crate) fn cancel(program_id: &Pubkey, acc: InstructionAccounts) -> ProgramResult {
+    msg!("Cancelling SPL token stream");
```
refresh_amounts()
as first line in cancel, withdraw, topup
to read the actual state from the chain (i.e. to see if additional funds have been deposited)
```
metadata.deposited_amount = escrow_tok_acc.amount + metadata.withdrawn_amount;
if (old_deposited_amount != metadata.deposited_amount) {
  //update fees
}
```

imprfekt committed:
just comments

imprfekt committed:
Merge branch 'feature/stream-config' of github.com:streamflow-finance/timelock-crate into feature/stream-config

imprfekt committed:
state: todo comments only

parazyd committed:
Make instruction functions public.

parazyd committed:
state: Add streamflow_fee_percent and partner_fee_percent to the metadata.

parazyd committed:
topup: Implement fee topup.

Commit:
merge stram-config to fees branch, tests WIP

parazyd committed:
create_stream: Implement separate accounts struct and checks.

parazyd committed:
WIP separate instruction account structs.

Commit:
provide control over bench accounts to tests, so variable number of accounts could be set up for each test

Commit:
merge test fascility changes

Commit:
adapt tests start_new

Commit:
added create stream tests

parazyd committed:
Finalize account struct separation.

Topup remains.

parazyd committed:
Remove some redundant reference borrows.

parazyd committed:
state: Remove InstructionAccounts struct.

Commit:
merge remote changes

Commit:
merge test changes

Commit:
add fees tests WIP

parazyd committed:
test-sdk: Remove ProcessInstructionWithContext dependency.

parazyd committed:
Update Cargo.lock

parazyd committed:
Cargo: Sort out dev dependencies for oracle program.

parazyd committed:
tests: Change references to accounts with new test-sdk and add fees concept.

parazyd committed:
withdraw: Handle requested amount from parameter.

Commit:
add withdraw authority verification

Commit:
merge remote changes

parazyd committed:
cargo +nightly fmt

parazyd committed:
create_stream: Add fee oracle to accounts.

parazyd committed:
utils: Implement function to check for external deposits.

parazyd committed:
withdraw: Use pubkeys from the metadata for the Invoker.

imprfekt committed:
create_stream: Check the deposited amount is more than period amount.

parazyd committed:
cancel_stream: Use metadata pubkeys for the Invoker.

parazyd committed:
transfer_stream: Rename Invoker variable.

parazyd committed:
topup_stream: Clean up spurious comment.

imprfekt committed:
license: update to Business License (for now)

parazyd committed:
utils: Add test unit for calculate_external_deposit()

imprfekt committed:
remove "_stream" & "_recipient" suffix from modules

imprfekt committed:
create: add doc comments

Commit:
setting up tests WIP

Commit:
merge remote changes

parazyd committed:
tests: Fix up partner oracle tests.

We add the shared object in the git tree.

imprfekt committed:
Merge branch 'fees' of github.com:streamflow-finance/timelock-crate into fees

imprfekt committed:
withdraw: doc comments

imprfekt committed:
withdraw: remove sender & sender_tokens

imprfekt committed:
added doc comments

imprfekt committed:
cargo +nightly fmt

imprfekt committed:
entrypoint: withdraw: remove sender, sender_tokens

imprfekt committed:
withdraw: accs: return to the original order

imprfekt committed:
remove total_amount field

imprfekt committed:
moved the constants & renamed variables

imprfekt committed:
errors, create: Added a new error

imprfekt committed:
create: Added explainer comments

imprfekt committed:
utils, create: Added a function for formatting

imprfekt committed:
rename TokenStreamData => Contract

imprfekt committed:
rename StreamInstruction => CreateStreamParams

imprfekt committed:
rename CreateStreamParams => CreateParams

imprfekt committed:
rename withdrawn_amount => amount_withdrawn

Want to have prefix amount_

imprfekt committed:
rename deposited_amount => amount_deposited

want to have prefix "amount_"

imprfekt committed:
rename deposited_amount => amount_deposited

imprfekt committed:
license update

imprfekt committed:
state: Added program version as a separate field

imprfekt committed:
state: is oke.

imprfekt committed:
state: just a pedantic naming convention

imprfekt committed:
todo: remove closable_at

@parazyd @nemanjab17

Commit:
test fix WIP

imprfekt committed:
create: Cast u8 to usize

imprfekt committed:
withdraw: can_withdraw takes "amount_requested" into consideration

imprfekt committed:
last_withdrawn_at: ignore withdrawing fees

Commit:
merge remote changes

Commit:
merge remote changes

Commit:
add deposit and sync balance methods; implement balance and deposit on withdraw and topup

imprfekt committed:
CreateParams: todo: remove end date as input parameter

imprfekt committed:
state: todo: Remove closable_at

imprfekt committed:
withdraw: close accounts after end date

imprfekt committed:
Merge branch 'fees' of github.com:streamflow-finance/timelock-crate into fees

Commit:
fix account data reference

Commit:
use util function for unpacking token acc

imprfekt committed:
TODO: end_time is calculated, not an input parameter

imprfekt committed:
TODO: - [ ] calculate if sender has enough funds to cover all the fees

imprfekt committed:
Merge branch 'fees' of github.com:streamflow-finance/timelock-crate into fees

imprfekt committed:
renamed `amount_deposited` => `net_amount_deposited`

imprfekt committed:
renamed `amount_deposited` => `net_amount_deposited`

imprfekt committed:
Removed todo

@parazyd zelimo da 100% uvek budu initialized pri withdraw, tako da sender placa za kreiranje acc.

imprfekt committed:
Use `amount_to_ui_amount` instead of custom `format` function

imprfekt committed:
removed unused import

Commit:
closable_at calculated on Stream create, moved closable_at on Create params

imprfekt committed:
use gross_amount

imprfekt committed:
create: comments and messages

Commit:
merge remote changes

Commit:
remove metadata.closable_at() method call  after move

Commit:
update closable_at whenever new deposit is identified; fix casts and build fails

Commit:
remove end time from create stream duration sanity check

Commit:
minor renames

Commit:
add lamports for rent check on create

imprfekt committed:
Create: removed todo and ran rustfmt

imprfekt committed:
Rename `closable_at` => `end_time`

imprfekt committed:
TODO: add can_topup

imprfekt committed:
todo: add can_topup

imprfekt committed:
withdraw: pass boolean instead of whole CreateParams struct

imprfekt committed:
withdraw: sanity check: updated error

imprfekt committed:
withdraw: update last_withdrawn_at only if withdrawn to recipient account

That is — ignore fee withdrawal. :)

imprfekt committed:
withdraw: account closing: move what's remaining to streamflow_token_treasury

Commit:
issue 19: implement can_topup, WIP

Commit:
merge remote changes

Commit:
move metadata save to a function

imprfekt committed:
code formatting: cargo +nightly fmt

imprfekt committed:
cancel: update canceled_at only in case of real cancel (not when now > `end_date` and the stream is just complete)

Commit:
make process.rs for common methods; extract close_escrow account method and call from cancel, withdraw

Commit:
merge remote changes

parazyd committed:
Use deterministic function for finding escrow accounts and validate them.

parazyd committed:
Migrate to latest escrow account derivation for all instructions.

Commit:
use metadata save on withdraw

parazyd committed:
state: Use ProgramResult for save_account_info.

parazyd committed:
process/close_escrow: Don't use references.

imprfekt committed:
grammar

parazyd committed:
Work through leftover TODO items.

parazyd committed:
Add new Javascript lib generator.

Currently we create the infos for instruction accounts.

parazyd committed:
state: Use a fixed-size u8 array for the stream name.

parazyd committed:
misc: Add (de)serialization.

parazyd committed:
js: Syntax fixes.

nemanjab17 commented:
Polish tasks:

- [ ] sender and recipient are the same - cancelation policy breaks

- [ ] validate amount per period > 0
